### PR TITLE
DRILL-4091: Support for additional gis operations in gis contrib module

### PR DESCRIPTION
**Pull request by ChrisSandison**
This based off of PR https://github.com/apache/drill/pull/258. I have attempted to address the comments from the PR left by @cgivre 

Some additional notes:

- in `STUnionAggregate.java`, there was a comment mentioning null handling, but I'm unable to find any null handling support when defining interfaces for aggregate functions.
- I have removed one of the failing tests from `TestGeometryFunctions.java`. The failing tests were attempting to test non-point geometry on functions that require it. While returning `Double.NaN` seems appropriate for the situation, I found two issues with testing this:

1. `DefaultSequelHandler.transform()` casts the results of the `Float8Holder` to `BigInteger` which does not have a representation for `NaN`. As a result, this is returning a `UserException` giving a system error, which would be the expected behaviour. However, I can't seem to test for this exception given the test builder that is used.
2. No other tests in the file are testing for invalid geometry. This is perhaps a larger issue.

### CODE REVIEW

**Comment by kkhatua** [comment]
@cgivre could you review this?

File: contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STCrosses.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.drill.exec.expr.DrillSimpleFunc;
+import org.apache.drill.exec.expr.annotations.FunctionTemplate;
+import org.apache.drill.exec.expr.annotations.Output;
+import org.apache.drill.exec.expr.annotations.Param;
+import org.apache.drill.exec.expr.holders.BitHolder;
+import org.apache.drill.exec.expr.holders.VarBinaryHolder;
+
+import io.netty.buffer.DrillBuf;
+
```
Comment (cgivre) [inline contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STCrosses.java]: Please include a sentence describing the functionality of this UDF.  Thanks!

File: contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STDifference.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import javax.inject.Inject;
+
+import org.apache.drill.exec.expr.DrillSimpleFunc;
+import org.apache.drill.exec.expr.annotations.FunctionTemplate;
+import org.apache.drill.exec.expr.annotations.Output;
+import org.apache.drill.exec.expr.annotations.Param;
+import org.apache.drill.exec.expr.holders.VarBinaryHolder;
+
+import io.netty.buffer.DrillBuf;
+
```
Comment (cgivre) [inline contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STDifference.java]: Please include a sentence describing the functionality of the UDF

File: contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STDisjoint.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.drill.exec.expr.DrillSimpleFunc;
+import org.apache.drill.exec.expr.annotations.FunctionTemplate;
+import org.apache.drill.exec.expr.annotations.Output;
+import org.apache.drill.exec.expr.annotations.Param;
+import org.apache.drill.exec.expr.holders.BitHolder;
+import org.apache.drill.exec.expr.holders.VarBinaryHolder;
+
+import io.netty.buffer.DrillBuf;
+
```
Comment (cgivre) [inline contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STDisjoint.java]: For all UDFs... I'd appreciate it if you could include a comment (one sentence will do and can be cut/pasted from PostGIS or wherever, describing what the function does.

**Comment by ChrisSandison** [comment]
@cgivre added comments

**Comment by cgivre** [comment]
@ChrisSandison This isn't a review comment, but @k255 was working on a format plugin for ESRI shape files.  Would you want to take a look and see if we could get that in as well?  I could see that being VERY useful once we get all this GIS functionality in.

**Comment by ChrisSandison** [comment]
@cgivre do you mean including that in this PR?

**Comment by cgivre** [comment]
If you want to include it on this PR that would be great, but I’m fine with including it in a separate PR as well.  I want to get this functionality into Drill asap, so it probably would be quicker to do a separate PR.

**Comment by ChrisSandison** [comment]
@cgivre I agree. Would you be able to open a JIRA ticket with any details of the work that @k255 was doing for it?

**Comment by cgivre** [comment]
@ChrisSandison Done.  Here's a link to the JIRA https://issues.apache.org/jira/browse/DRILL-6319

**Comment by ChrisSandison** [comment]
@cgivre I'm also needing to use ST_DWithin (https://postgis.net/docs/ST_DWithin.html). I'm going to add it to this PR.

**Comment by cgivre** [comment]
LGTM +1

One thing, which I suspect may be a problem with my environment, not this PR, but when I try to build the module with the tests, I get the following errors:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:2.6:single (source-release-assembly) on project drill-gis: Error reading assemblies: Error locating assembly descriptor: src/main/resources/assemblies/source-assembly.xml
[ERROR] 
[ERROR] [1] [INFO] Searching for file location: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml
[ERROR] 
[ERROR] [2] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.
[ERROR] 
[ERROR] [3] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:2.6:single (source-release-assembly) on project drill-gis: Error reading assemblies: Error locating assembly descriptor: src/main/resources/assemblies/source-assembly.xml

[1] [INFO] Searching for file location: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml

[2] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.

[3] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:497)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error reading assemblies: Error locating assembly descriptor: src/main/resources/assemblies/source-assembly.xml

[1] [INFO] Searching for file location: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml

[2] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.

[3] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.
    at org.apache.maven.plugin.assembly.mojos.AbstractAssemblyMojo.execute (AbstractAssemblyMojo.java:453)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:497)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.assembly.io.AssemblyReadException: Error locating assembly descriptor: src/main/resources/assemblies/source-assembly.xml

[1] [INFO] Searching for file location: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml

[2] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.

[3] [INFO] File: /Users/cgivre/github/drill-dev/drill-gis/drill/contrib/gis/src/main/resources/assemblies/source-assembly.xml does not exist.
    at org.apache.maven.plugin.assembly.io.DefaultAssemblyReader.addAssemblyFromDescriptor (DefaultAssemblyReader.java:293)
    at org.apache.maven.plugin.assembly.io.DefaultAssemblyReader.readAssemblies (DefaultAssemblyReader.java:109)
    at org.apache.maven.plugin.assembly.mojos.AbstractAssemblyMojo.execute (AbstractAssemblyMojo.java:449)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:497)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Charles-MacBook-Pro:gis cgivre$ 
```

**Comment by ChrisSandison** [comment]
@cgivre I had some issues between this and a previous branch I was working on, but I fixed it by upgrade the base image that I was building it on. I ended up building it on maven:3.5.3-jdk-8

**Comment by cgivre** [comment]
In that case,  LGTM +1 
Thank you for your work on this!.  Are you going to take a look at the ESRI format plugin?

**Comment by cgivre** [comment]
@arina-ielchiieva Should I do the commit or would you like to do that.

**Comment by cgivre** [comment]
Thanks @arina-ielchiieva 

@ChrisSandison Can you please:
1.  Squash commits into one
2.  Apply the Drill coding styles 
3.  Fix the license headers

And we'll get it committed!  Thanks!

**Comment by ChrisSandison** [comment]
@cgivre updated and squashed

**Comment by cgivre** [comment]
@ChrisSandison, 
The PR is failing the TravisCI build due to unused import statements:

```
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STYFunc.java:29:8: Unused import - com.esri.core.geometry.Geometry.Type.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STYFunc.java:30:8: Unused import - com.esri.core.geometry.ogc.OGCPoint.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STXFunc.java:20:8: Unused import - java.sql.Types.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STXFunc.java:31:8: Unused import - com.esri.core.geometry.Geometry.Type.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STXFunc.java:32:8: Unused import - com.esri.core.geometry.ogc.OGCPoint.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STUnionAggregate.java:33:8: Unused import - com.esri.core.geometry.SpatialReference.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STGeomFromText.java:22:8: Unused import - com.esri.core.geometry.ogc.OGCGeometry.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STGeomFromText.java:32:8: Unused import - java.nio.ByteBuffer.
/home/travis/build/apache/drill/contrib/gis/src/main/java/org/apache/drill/exec/expr/fn/impl/gis/STTransform.java:28:8: Unused import - org.apache.drill.exec.expr.holders.NullableVarCharHolder.

```

Could you please remove these, squash the commits and I think we're done.

**Comment by cgivre** [comment]
+1

**Comment by cgivre** [comment]
@ChrisSandison 
Thank you for your contribution to Drill!

**Comment by ChrisSandison** [comment]
@parthchandra @cgivre why has this been closed?

**Comment by parthchandra** [comment]
It's committed into master.

**Pull request closed without merging**